asio/include/asio/fiber.hpp

```diff
@@ -0,0 +1,1553 @@
+
+  //// move construct a basic_fiber_context from anoter context. Can convert the executor type.
+  template<typename Executor2>
+  basic_fiber_context(
+          basic_fiber_context<Executor2>  && other,
+          typename constraint<std::is_convertible_v<Executor2, Executor>>::type = 0) noexcept
+          : handle_(other.handle_), fiber_(std::move(other.fiber_)), executor_(std::move(other.get_executor()))
+  {
+  }
+#if !defined(GENERATING_DOCUMENTATION)
+
+  basic_fiber_context(detail::fiber_handle &handle,
+                      boost::context::fiber && fiber,
+                      executor_type exec) : handle_(&handle), fiber_(&fiber), executor_(std::move(exec))
+  {
+  }
+
+#endif
+
+  /// Get the executor of the fiber.
+  executor_type get_executor() const {return executor_;}
+
+  /// Get the cancellation state of the fiber.
+  cancellation_state get_cancellation_state() const ASIO_NOEXCEPT
+  {
+    return handle_->cancel_state;
+  }
+
+  /// Reset the cancellation state of the fiber.
+  void reset_cancellation_state()
+  {
+    handle_->cancel_state = cancellation_state(handle_->cancel_slot);
+  }
+
+  /// Reset the cancellation state of the fiber and insert a filter.
+  template <typename Filter>
+  void reset_cancellation_state(ASIO_MOVE_ARG(Filter) filter)
+  {
+    handle_->cancel_state = cancellation_state(handle_->cancel_slot, ASIO_MOVE_CAST(Filter)(filter));
+  }
+
+  /// Reset the cancellation state of the fiber and insert in and out filters.
+  template <typename InFilter, typename OutFilter>
+  void reset_cancellation_state(ASIO_MOVE_ARG(InFilter) in_filter,
+                                ASIO_MOVE_ARG(OutFilter) out_filter)
+  {
+    handle_->cancel_state = cancellation_state(handle_->cancel_slot,
+                                              ASIO_MOVE_CAST(InFilter)(in_filter),
+                                              ASIO_MOVE_CAST(OutFilter)(out_filter));
+  }
+
+ public:
+
+  template<typename H, ASIO_COMPLETION_SIGNATURES_TPARAMS>
+  friend class async_result;
+
+  detail::fiber_handle *handle_;
+  // innter fiber
+  boost::context::fiber *fiber_;
+  executor_type executor_;
+
+};
+
+/// The fiber context type with the default executor.
+typedef basic_fiber_context<> fiber_context;
+
+namespace detail
+{
+
+template<typename Executor>
+struct initiate_fiber
+{
+
+  template<typename CompletionHandler, typename Func>
+  void operator()(CompletionHandler && completion_handler, Func && function, void(*)()) const
+  {
+    auto alloc = get_associated_executor(completion_handler, get_executor());
+
+    struct fiber_runner : fiber_handle
+    {
+      Executor executor;
+      Func func;
+      CompletionHandler handler;
+
+      fiber_runner(cancellation_slot slot, Executor executor, Func && func, CompletionHandler && handler)
+          : fiber_handle(slot), executor(std::move(executor)),
+            func(std::forward<Func>(func)), handler(std::forward<CompletionHandler>(handler)) {}
+
+      auto complete()
```

Comment:
`auto` return is C++17?

Reply:
I don't like the look of this one single bit:
```
executor(std::move(executor))
```

Reply:
It most likely be under a problem with -Wshadow (includes some other ARGS)